chain/actors/builtin/init/state.go.template

```diff
@@ -119,10 +128,49 @@ func (s *state{{.v}}) SetAddressMap(mcid cid.Cid) error {
 	return nil
 }
 
+func (s *state{{.v}}) GetState() interface{} {
+	return &s.State
+}
+
+func (s *state{{.v}}) ActorKey() string {
+    return actors.InitKey
+}
+
+func (s *state{{.v}}) ActorVersion() actors.Version {
+    return actors.Version{{.v}}
+}
+
+func (s *state{{.v}}) Code() cid.Cid {
+	code, ok := actors.GetActorCodeID(s.ActorVersion(), s.ActorKey())
+	if !ok {
+		panic(fmt.Errorf("didn't find actor %v code id for actor version %d", s.ActorKey(), s.ActorVersion()))
+	}
+
+	return code
+}
+
 func (s *state{{.v}}) AddressMap() (adt.Map, error) {
 	return adt{{.v}}.AsMap(s.store, s.State.AddressMap{{if (ge .v 3)}}, builtin{{.v}}.DefaultHamtBitwidth{{end}})
 }
 
-func (s *state{{.v}}) GetState() interface{} {
-	return &s.State
-}
\ No newline at end of file
+func (s *state{{.v}}) AddressMapBitWidth() int {
```

Comment:
Ideas on how to make the implementation more compact?
in `chain/actors/adt/adt.go` I've defined:
```go
type MapHashFunc func([]byte) []byte

type MapOpts interface {
	BitWidth() int
	HashFunc() MapHashFunc
}
```
implementing the interface then looks something like:
```go
type mapOpts struct {
	hashFunc adt.MapHashFunc
	bitWidth func() int
}

func (m mapOpts) HashFunc() adt.MapHashFunc { return m.hashFunc }
func (m mapOpts) BitWidth() int             { return m.bitWidth() }

func (s *state2) AddressMap() (adt.Map, adt.MapOpts, error) {
	m, err := adt2.AsMap(s.store, s.State.AddressMap)
	if err != nil {
		return nil, nil, err
	}
	t := struct{ mapOpts }{}
	t.bitWidth = func() int { return 5 }
	t.hashFunc = func(input []byte) []byte {
		res := sha256.Sum256(input)
		return res[:]
	}
	return m, t, nil
}
```

I'd like to avoid redefining `mapOpts` everywhere it's needed to satisfy the `adt.MapOpts` interface; couple ideas:
1) make MapOpts a structure and drop the whole interface thing, may be annoying to work with later if we suspect there to be more options
2) keep the MapOpts interface and move the `mapOpts` structure somewhere in the `chain/actors` package
3) don't return a structure or an interface and instead define `AddressMap() (adt.Map, int, ad.MapHashFunc, error)`

Reply:
I didn't look too deeply at this but making MapOpts a struct seems great.  We can always come back and generalize and extend later but for now lets keep it simple

Reply:
Great, I will leave this as is then.